WIP: Add compat tree rebuild support

f-trivino opened the pull request:
For different reasons, the compat tree might get corrupted and some entries would be missing.

This PR implements a new IPA server command: "ipa server-compat-tree-refresh" that allows the admin to rebuild the compat tree without having to restart the LDAP server.

Fixes: https://pagure.io/freeipa/issue/9219

Todo:
- [x] Add new IPA command "ipa server-compat-tree-refresh"
- [x] Create dn container during install/upgrade as well as the proper ACI allowing to create the task below the new container.
- [x] Design page
- [x] Merge slapi-nis PR https://pagure.io/slapi-nis/pull-request/46
- [x] Spec file changes to make sure slapi-nis supports the refresh of the compat tree
- [ ] Implement XMLRPC tests

f-trivino commented:
Tested in Fedora 36 using COPR https://copr.fedorainfracloud.org/coprs/ftrivino/slapi-nis/
```
[root@master fedora]# kinit admin
Password for admin@IPA.TEST: 
[root@master fedora]# ipa compat-tree-rebuild
---------------------------
Refresh maps task finished.
---------------------------
[root@master fedora]# ipa compat-tree-rebuild --no-wait
----------------------------------------------
Schema compatibility tree rebuild task started
----------------------------------------------
  Task DN: cn=1d2c1511-e1a8-4517-b77f-58e1af27b261,cn=Schema compatibility refresh task,cn=tasks,cn=config

Aug 07 21:34:19 master.ipa.test ns-slapd[18307]: [07/Aug/2022:21:34:19.947354818 +0000] - INFO - schema-compat-plugin - refresh_thread --> refcount incremented.
Aug 07 21:34:19 master.ipa.test ns-slapd[18307]: [07/Aug/2022:21:34:19.955713484 +0000] - INFO - schema-compat-plugin - backend_shr_refresh_thread - Refresh task starts
Aug 07 21:34:19 master.ipa.test ns-slapd[18307]: [07/Aug/2022:21:34:19.960061167 +0000] - INFO - schema-compat-plugin - backend_shr_refresh_thread - Refresh maps starting soon.
Aug 07 21:34:19 master.ipa.test ns-slapd[18307]: [07/Aug/2022:21:34:19.962382511 +0000] - ERR - schema-compat-plugin - schema-compat-plugin tree scan will start in about 5 seconds!
Aug 07 21:34:24 master.ipa.test ns-slapd[18307]: [07/Aug/2022:21:34:24.968694560 +0000] - ERR - schema-compat-plugin - warning: no entries set up under ou=sudoers,dc=ipa,dc=test
Aug 07 21:34:24 master.ipa.test ns-slapd[18307]: [07/Aug/2022:21:34:24.974426129 +0000] - ERR - schema-compat-plugin - warning: no entries set up under cn=ng, cn=compat,dc=ipa,dc=test
Aug 07 21:34:24 master.ipa.test ns-slapd[18307]: [07/Aug/2022:21:34:24.978003298 +0000] - ERR - schema-compat-plugin - warning: no entries set up under cn=computers, cn=compat,dc=ipa,dc=test
Aug 07 21:34:24 master.ipa.test ns-slapd[18307]: [07/Aug/2022:21:34:24.980120298 +0000] - ERR - schema-compat-plugin - Finished plugin initialization.
Aug 07 21:34:25 master.ipa.test ns-slapd[18307]: [07/Aug/2022:21:34:25.965034260 +0000] - INFO - schema-compat-plugin - backend_shr_refresh_thread - Refresh maps task finished.
Aug 07 21:34:25 master.ipa.test ns-slapd[18307]: [07/Aug/2022:21:34:25.968752714 +0000] - INFO - schema-compat-plugin - backend_shr_refresh_thread <-- refcount decremented.
```

tbordaz commented on doc/designs/compat-tree-rebuild.md:
```diff
@@ -0,0 +1,88 @@
+    objectClass: extensibleObject
+    cn: Schema compatibility refresh task
+
+Then, the admin can issue the command `ipa automember-rebuild`. This call internally
```
Do you mean 'compat-tree-rebuild' ?

f-trivino replied:
yep, thanks @tbordaz

abbra commented:
Thank you for this PR.

* I'd like to see this moved under `ipa server-*` set of commands. We should try to reuse existing namespaces rather than new ones. May be something like `ipa server-compat-tree-refresh` would be a better command?
* Only administrators can create new tasks for the compat tree rebuild, the plugin code doesn't handle it and also it doesn't handle any fallouts from the failures to add an LDAP entry for the task. Please see implementation of `config_mod._enable_sid()` method for an example how it was done.
* You need also to detect that the server actually has the compat tree enabled and it supports tree rebuilds.
* what is the `compat_task` object's purpose? You are not using it really anywhere in the code so you cannot look up a result of the task operation. May be add a separate `ipa server-compat-tree-task-show` command that accepts the name of the task and looks it up in the list by itself. Then we should not be publish the DN directly.

f-trivino commented:
Thanks @abbra for the review, 

--- comments inline ---

> Thank you for this PR.
> 
>     * I'd like to see this moved under `ipa server-*` set of commands. We should try to reuse existing namespaces rather than new ones. May be something like `ipa server-compat-tree-refresh` would be a better command?

I agree, I moved the content to the ipa server plugin, so the new command is `ipa server-compat-tree-refresh`

> 
>     * Only administrators can create new tasks for the compat tree rebuild, the plugin code doesn't handle it
I'm directly adding permission, privilege, and ACI into install/upgrades/80-schema_compat.update LDIF instead of `managed permission` as the later is not always easily applicable, I guess it is OK to do so for this simple use case.

> and also it doesn't handle any fallouts from the failures to add an LDAP entry for the task. Please see implementation of `config_mod._enable_sid()` method for an example how it was done.

Added.

>     * You need also to detect that the server actually has the compat tree enabled and it supports tree rebuilds.

I added slapi-nis-0.60.0-1 in the specfile for both fedora and rhel.

> 
>     * what is the `compat_task` object's purpose? You are not using it really anywhere in the code so you cannot look up a result of the task operation. May be add a separate `ipa server-compat-tree-task-show` command that accepts the name of the task and looks it up in the list by itself. Then we should not be publish the DN directly.

my bad, compat_task was not supposed to be there. Removed. Ref `ipa server-compat-tree-task-show` I propose to go for a more common ipa command, something like `ipa task-show` that can be useful for more use cases (e.g. automount), stuff from "cn=tasks,cn=config" and could be implemented later in a separate PR.

I started to test the implementation using ftrivino/slapi-nis COPR, and plan to add an additional XMLRPC test commit to this PR.

https://copr.fedorainfracloud.org/coprs/ftrivino/slapi-nis/builds/
